### PR TITLE
[SPARK-25021][K8S] Add spark.executor.pyspark.memory limit for K8S

### PR DESCRIPTION
**Pull request by ifilonenko**
## What changes were proposed in this pull request?

Add spark.executor.pyspark.memory limit for K8S

## How was this patch tested?

Unit and Integration tests

### CODE REVIEW

**Comment by ifilonenko** [comment]
@rdblue @holdenk for review. This contains both unit and integration tests that verify [SPARK-25004] for K8S

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/SecretsTestsSuite.scala [hunk 1]
```diff
@@ -53,6 +53,7 @@ private[spark] trait SecretsTestsSuite { k8sSuite: KubernetesSuite =>
       .delete()
   }
 
+  // TODO: [SPARK-25291] This test is flaky with regards to memory of executors
```
Comment (ifilonenko) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/SecretsTestsSuite.scala]: @mccheah This test periodically fails on setting proper memory for executors on this specific test. I have filed a JIRA: SPARK-25291

File: examples/src/main/python/worker_memory_check.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+#
```
Comment (holdenk) [inline examples/src/main/python/worker_memory_check.py]: Is examples the right place for this?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 3]
```diff
@@ -225,6 +225,13 @@ private[spark] object Config extends Logging {
         "Ensure that major Python version is either Python2 or Python3")
       .createWithDefault("2")
 
+  val APP_RESOURCE_TYPE =
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala]: Why this instead of the bools? What about folks who want to make a pipeline which is both R and Python?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala [hunk 4]
```diff
@@ -51,6 +51,7 @@ private[spark] class BasicDriverFeatureStep(
     .get(DRIVER_MEMORY_OVERHEAD)
     .getOrElse(math.max((conf.get(MEMORY_OVERHEAD_FACTOR) * driverMemoryMiB).toInt,
       MEMORY_OVERHEAD_MIN_MIB))
+  // TODO: Have memory limit checks on driverMemory
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala]: Can we file a JIRA as well and include it in the comment.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala [hunk 5]
```diff
@@ -58,6 +58,15 @@ private[spark] class BasicExecutorFeatureStep(
       (kubernetesConf.get(MEMORY_OVERHEAD_FACTOR) * executorMemoryMiB).toInt,
       MEMORY_OVERHEAD_MIN_MIB))
   private val executorMemoryWithOverhead = executorMemoryMiB + memoryOverheadMiB
+  // TODO: Have memory limit checks on executorMemory
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala]: Same as other TODO

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala [hunk 6]
```diff
@@ -58,6 +58,15 @@ private[spark] class BasicExecutorFeatureStep(
...
+  private val executorMemoryTotal = kubernetesConf.sparkConf
+    .getOption(APP_RESOURCE_TYPE.key).map{ res =>
+      val additionalPySparkMemory = if (res == "python") {
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala]: So this means that we couldn't turn this on in a mixed language pipeline even if the pipeline author did some fun hacks (since we might need to support Scala/Python/R all at the same time). Is there a reason we did this instead of the YARN approach of isPython ?

Also I'd write this as a case statement instead perhaps style wise, what do you think?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/bindings/JavaDriverFeatureStep.scala [hunk 7]
```diff
@@ -38,7 +39,8 @@ private[spark] class JavaDriverFeatureStep(
       .build()
     SparkPod(pod.pod, withDriverArgs)
   }
-  override def getAdditionalPodSystemProperties(): Map[String, String] = Map.empty
+  override def getAdditionalPodSystemProperties(): Map[String, String] =
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/bindings/JavaDriverFeatureStep.scala]: So is this going to override it if the user wants to set a different type even if they are ostensibly running a JVM pod (e.g. mixed language pipelines).

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/bindings/PythonDriverFeatureStep.scala [hunk 8]
```diff
@@ -67,7 +68,8 @@ private[spark] class PythonDriverFeatureStep(
 
     SparkPod(pod.pod, withPythonPrimaryContainer)
   }
-  override def getAdditionalPodSystemProperties(): Map[String, String] = Map.empty
+  override def getAdditionalPodSystemProperties(): Map[String, String] =
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/bindings/PythonDriverFeatureStep.scala]: Same as above (just curious if this overrides user params or precedence rules here). Also add a comment of the precedence rules.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/bindings/RDriverFeatureStep.scala [hunk 9]
```diff
@@ -53,7 +54,8 @@ private[spark] class RDriverFeatureStep(
 
     SparkPod(pod.pod, withRPrimaryContainer)
   }
-  override def getAdditionalPodSystemProperties(): Map[String, String] = Map.empty
+  override def getAdditionalPodSystemProperties(): Map[String, String] =
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/bindings/RDriverFeatureStep.scala]: Same

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile [hunk 10]
```diff
@@ -19,10 +19,10 @@ ARG base_img
 FROM $base_img
 WORKDIR /
 RUN mkdir ${SPARK_HOME}/R
-COPY R ${SPARK_HOME}/R
 
 RUN apk add --no-cache R R-dev
 
+COPY R ${SPARK_HOME}/R
```
Comment (holdenk) [inline resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile]: Is this change intentional?

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/python/Dockerfile [hunk 11]
```diff
@@ -33,6 +32,7 @@ RUN apk add --no-cache python && \
     # Removed the .cache to save space
     rm -r /root/.cache
 
+COPY python/lib ${SPARK_HOME}/python/lib
```
Comment (holdenk) [inline resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/python/Dockerfile]: Same, is this change intentional

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala [hunk 12]
```diff
@@ -72,12 +72,33 @@ private[spark] trait PythonTestsSuite { k8sSuite: KubernetesSuite =>
       isJVM = false,
       pyFiles = Some(PYSPARK_CONTAINER_TESTS))
   }
+
+  test("Run PySpark with memory customization", k8sTestTag) {
+    sparkAppConf
+      .set("spark.kubernetes.container.image", s"${getTestImageRepo}/spark-py:${getTestImageTag}")
```
Comment (holdenk) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala]: nit: Mildly confused why there's so many sets here (like the image, etc.) maybe make more sense in a shared test setup func?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala [hunk 13]
```diff
@@ -72,12 +72,33 @@ private[spark] trait PythonTestsSuite { k8sSuite: KubernetesSuite =>
...
+      .set("spark.kubernetes.pyspark.pythonVersion", "3")
+      .set("spark.kubernetes.memoryOverheadFactor", s"$memOverheadConstant")
+      .set("spark.executor.pyspark.memory", s"${additionalMemory}m")
+      .set("spark.python.worker.reuse", "false")
```
Comment (holdenk) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala]: Why is worker reuse being set to false?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala [hunk 14]
```diff
@@ -72,12 +72,33 @@ private[spark] trait PythonTestsSuite { k8sSuite: KubernetesSuite =>
       isJVM = false,
       pyFiles = Some(PYSPARK_CONTAINER_TESTS))
   }
+
+  test("Run PySpark with memory customization", k8sTestTag) {
+    sparkAppConf
+      .set("spark.kubernetes.container.image", s"${getTestImageRepo}/spark-py:${getTestImageTag}")
+      .set("spark.kubernetes.pyspark.pythonVersion", "3")
+      .set("spark.kubernetes.memoryOverheadFactor", s"$memOverheadConstant")
```
Comment (holdenk) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala]: Do we expect people who configure the rlimit advanced feature to also set the memoryOverheadConstant to a different value? If so we should call it out in the docs. (note: I think it would make sense for folks to set this to a lower value so I _think_ this would be the expected behaviour and we should document but open to suggestions)

**Review by holdenk — COMMENTED** [review]
Did a quick first pass over this PR, really excited to have this support in K8s as well as yarn, but I have some questions especially around mixed language pipelines. 

Really excited the K8s integration tests are now integrated.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala (hunk 3 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala]: The reason for this is because we are already running binding steps that configure the driver based on the app resource. I thought it might as well pass the config down into the executors upon doing that binding bootstrap step. 

Currently, we don't have any docker files that handle mixed pipelines so such configurations should be addressed in a followup-PR, imo. But I am open to suggestions (that are testable).

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala (hunk 4 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala]: I wanted to get an opinion from people (@mccheah) into whether or not we wanted to let the K8S api to handle memory limits (via ResourceQuota limit errors) or whether we wanted to catch it in a Spark exception (if we were to include a configuration for memory limits)

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala (hunk 6 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala]: Well, if there is a mixed pipeline, the binding steps would need to be reconfigured to include mixed pipelines. But currently, language is only determined based on the `appResource`

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile (hunk 10 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile]: This makes the docker build not run the `R` and `R-dev` installations each time an update to the jar is made. This is a minor change that helps with dev :)

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/python/Dockerfile (hunk 11 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/python/Dockerfile]: Same as the R change above ^^

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala (hunk 14 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala]: I can add to docs, sure!

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala (hunk 13 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala]: in reference to https://github.com/apache/spark/pull/21977#issuecomment-411130872

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala (hunk 14 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala]: This is already included in the docs of `docs/running-on-kubernetes.md`

File: examples/src/main/python/worker_memory_check.py (hunk 2 above)
Comment (ifilonenko) [reply examples/src/main/python/worker_memory_check.py]: Easiest place to put to be trivially picked up by integration tests, as I did with pyfiles, but I am open to recommendations.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala (hunk 4 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala]: Hm can you elaborate here? We already set the driver memory limit in this step based on the overhead.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/bindings/JavaDriverFeatureStep.scala (hunk 7 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/bindings/JavaDriverFeatureStep.scala]: Again, not sure what our direction is going to be with respect to mixed pipelines throughout the cluster manager sections - if we should be supporting it in a first class way then perhaps we file a JIRA and we can discuss how the submission client should be refactored to support that.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala (hunk 12 above)
Comment (mccheah) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala]: Some of this stuff can be factored out I think, we just haven't done so yet. I wouldn't block a merge of this on such a refactor, but this entire test class could probably use some cleanup with respect to how the code is structured.

**Comment by rdblue** [comment]
Looks fine to me, but I'm not familiar enough with the K8S code to have much of an opinion.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala (hunk 4 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala]: Valid point. These are not necessary

**Comment by ifilonenko** [comment]
@holdenk and @mccheah any other comments before merge?

**Comment by mccheah** [comment]
https://issues.apache.org/jira/browse/SPARK-25291 looks like a real issue with the way the tests are written. So we don't necessarily want to ignore it for this patch, but we're still thinking about it.

**Comment by ifilonenko** [comment]
@mccheah as such, that should be a followup PR and this should be good to merge as long as  @holdenk gives a LGTM

File: examples/src/main/python/worker_memory_check.py (hunk 2 above)
Comment (felixcheung) [reply examples/src/main/python/worker_memory_check.py]: shouldn't this be in python tests (and get it to run only certain cluster manager)

Comment (ifilonenko) [reply examples/src/main/python/worker_memory_check.py]: That might be a good place for it. But the reason for this being in examples is that the integration tests can access this locally and we can see the success in the Jenkins environment. The K8s integration test suite does not run python run-tests.py which means that this test would therefore not be part of the PRB.

**Comment by ifilonenko** [comment]
@felixcheung @holdenk I have moved the PySpark example files to a more appropriate location. Any other comments before merge?

**Comment by mccheah** [comment]
Yeah this looks ok to me, would like a +1 from @felixcheung and @holdenk.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala (hunk 3 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala]: Yeah lets do something in a follow up after 2.4

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/bindings/JavaDriverFeatureStep.scala (hunk 7 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/bindings/JavaDriverFeatureStep.scala]: filed SPARK-25373 - Support mixed language pipelines on Spark on K8s

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala [hunk 15]
```diff
@@ -72,12 +74,33 @@ private[spark] trait PythonTestsSuite { k8sSuite: KubernetesSuite =>
...
       pyFiles = Some(PYSPARK_CONTAINER_TESTS))
   }
+
+  test("Run PySpark with memory customization", k8sTestTag) {
+    sparkAppConf
+      .set("spark.kubernetes.container.image", pySparkDockerImage)
+      .set("spark.kubernetes.pyspark.pythonVersion", "3")
+      .set("spark.kubernetes.memoryOverheadFactor", s"$memOverheadConstant")
+      .set("spark.executor.pyspark.memory", s"${additionalMemory}m")
+      .set("spark.python.worker.reuse", "false")
```
Comment (holdenk) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/PythonTestsSuite.scala]: I don't believe this should be set. Worker reuse is on by default in most systems so not sure if this test depends on worker reuse being false. As per @rdblue's investigation this shouldn't impact this code path (and if it does we need to re-open that investigation).

**Review by holdenk — COMMENTED** [review]
Two minor nits, I think we should get rid of the set to false but otherwise LGTM since K8s folks have already signed off on that part.

**Comment by holdenk** [comment]
LGTM pending jenkins, see previous comments for details.

**Comment by ifilonenko** [comment]
@holdenk or @mccheah for merge
